securitas/controller/authentication.py

```diff
@@ -1,43 +1,55 @@
-from flask import flash, redirect, request, session, url_for
+from flask import flash, redirect, session, url_for, render_template
 import python_freeipa
 
 from securitas import app
-from securitas.security.ipa import maybe_ipa_login
-from securitas.utility import with_ipa
+from securitas.form.login_user import LoginUserForm
+from securitas.security.ipa import maybe_ipa_login, maybe_ipa_session
 
 
 @app.route('/logout')
-@with_ipa(app, session)
-def logout(ipa):
+def logout():
+    """Log the user out."""
+    # Don't use the with_ipa() decorator, otherwise anonymous users visiting this endpoint will be
+    # asked to login to then be logged out.
+    ipa = maybe_ipa_session(app, session)
     if ipa:
         ipa.logout()
     session.clear()
     return redirect(url_for('root'))
 
 
-@app.route('/login', methods=['POST'])
+@app.route('/login', methods=['GET', 'POST'])
 def login():
-    username = request.form.get('username')
-    password = request.form.get('password')
-    if not username or not password:
-        flash('Please provide both a username and a password.', 'red')
-        return redirect(url_for('root'))
-
-    try:
-        # This call will set the cookie itself, we don't have to.
-        ipa = maybe_ipa_login(app, session, username, password)
-    except python_freeipa.exceptions.PasswordExpired:
-        flash('Password expired. Please reset it.', 'red')
-        return redirect(url_for('password_reset'))
-    except python_freeipa.exceptions.Unauthorized as e:
-        flash(str(e), 'red')
-        return redirect(url_for('root'))
-
-    if ipa:
-        flash('Welcome, %s!' % username, 'green')
-        return redirect(url_for('user', username=username))
-
-    # If we made it here, we hit something weird not caught above. We didn't
-    # bomb out, but we don't have IPA creds, either. Boot us back to /.
-    flash('Could not log in to the IPA server.', 'red')
-    return redirect(url_for('root'))
+    form = LoginUserForm()
+    if form.validate_on_submit():
+        username = form.username.data
+        password = form.password.data
+        try:
+            # This call will set the cookie itself, we don't have to.
+            ipa = maybe_ipa_login(app, session, username, password)
+        except python_freeipa.exceptions.PasswordExpired:
+            flash('Password expired. Please reset it.', 'red')
+            return redirect(url_for('password_reset'))
+        except python_freeipa.exceptions.Unauthorized as e:
+            form.errors['non_field_errors'] = [e.message]
+        except python_freeipa.exceptions.FreeIPAError as e:
+            # If we made it here, we hit something weird not caught above. We didn't
+            # bomb out, but we don't have IPA creds, either. Boot us back to /.
+            app.logger.error(
```

Comment:
I *think* this is okay because the IPA server should never include passwords in the response, but we should be mindful to make sure nothing we log will ever include a password.

Reply:
Oh good point, I didn't think of that. Hopefully they don't put passwords in the `message` attribute of their exceptions, though.